tests/conftest.py

```diff
@@ -0,0 +1,107 @@
+import tempfile
+import pathlib
+from time import time
+from os import environ
+from typing import Callable, Tuple
+
+import pytest
+import alembic.config
+from fastapi import FastAPI
+from fastapi.testclient import TestClient
+
+from tests.certificates import PRIVATE_KEY, PUBLIC_KEY
+
+
+@pytest.fixture(scope="session")
+def test_certificates() -> Tuple[str, str]:
+    return PRIVATE_KEY, PUBLIC_KEY
+
+
+@pytest.fixture(scope="session")
+def find_free_port() -> Callable:
+    """
+    Returns a factory that finds the next free port that is available on the OS
+    This is a bit of a hack, it does this by creating a new socket, and calling
+    bind with the 0 port. The operating system will assign a brand new port,
+    which we can find out using getsockname(). Once we have the new port
+    information we close the socket thereby returning it to the free pool.
+    This means it is technically possible for this function to return the same
+    port twice (for example if run in very quick succession), however operating
+    systems return a random port number in the default range (1024 - 65535),
+    and it is highly unlikely for two processes to get the same port number.
+    In other words, it is possible to flake, but incredibly unlikely.
+    """
+
+    def _find_free_port() -> int:
+        import socket
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.bind(("0.0.0.0", 0))
+        portnum = s.getsockname()[1]
+        s.close()
+
+        return portnum
+
+    return _find_free_port
+
+
+@pytest.fixture(scope="session", autouse=True)
+def testing_environment(
+    test_certificates: Tuple[str, str], find_free_port: Callable
+) -> None:
+    location = pathlib.Path(tempfile.gettempdir())
+    tstamp = int(time())
```

Comment:
Can we add a few random characters on the end of this? Might help reduce collisions.

Ye olde random letters:
```
''.join(random.choices(string.ascii_uppercase +
                             string.digits, k = 6))
```

Reply:
This is scoped to the pytest session, which is equivalent to the whole run on the test suite. As in, it is only run at the start, yielded during the suite, then cleans up. It would take two pytest instances to be trigged within a second for a collision to happen. Granted, random characters alone would protect against even that.